compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,453 @@
+  return true;
+}
+
+/**
+ * @brief update _shape based on 'Reshape' shape value
+ *
+ * @example Let's assume Reshape(shape=1, 448, 49) is given to [before] _shape.
+ *
+ *  [before]  _shape :
+ *              - value :   (1)   (448)  (7)   (7)
+ *              - tags  :   (-)   (2)    (0)   (1)
+ *
+ *  [after]  _shape :
+ *              - value :   (1)   (448)  (49)
+ *              - tags  :   (-)   (2)    (0, 1)
+ */
+template <loco::DataType ReshapeType>
+bool TaggedShapeAnalyzer::analyze_reshape(const luci::CircleReshape *reshape_node)
+{
+  const luci::CircleConst *shape_node = loco::must_cast<luci::CircleConst *>(reshape_node->shape());
+  assert(shape_node->dtype() == ReshapeType);
+
+  // At least one element must be in reshape's output-tensor.
+  if (shape_node->size<ReshapeType>() <= 0)
+    return false;
+
+  // Create new_shape based on reshape_node/shape
+  Shape new_shape;
+  for (uint32_t i = 0; i < shape_node->size<ReshapeType>(); i++)
+  {
+    TaggedShapeAnalyzer::Dim dim;
+    dim.value = shape_node->at<ReshapeType>(i);
+    new_shape.push_back(dim);
+  }
+
+  // indexing for _shape [old_shape_start_idx, old_shape_end_idx)
+  uint32_t old_shape_start_idx = 0;
+  uint32_t old_shape_end_idx = 1;
+  uint32_t old_shape_product = _shape[old_shape_start_idx].value;
+
+  auto expand_range = [&]() -> bool {
+    if (old_shape_end_idx >= _shape.size())
+      return false;
+
+    old_shape_product *= _shape[old_shape_end_idx].value;
+    old_shape_end_idx++;
+    return true;
+  };
+
+  auto move_to_next_range = [&]() -> bool {
+    if (old_shape_end_idx > _shape.size())
+      return false;
+
+    old_shape_start_idx = old_shape_end_idx;
+    old_shape_end_idx++;
+    old_shape_product = _shape[old_shape_start_idx].value;
+    return true;
+  };
+
+  // Add tags from '_shape' to the 'new_shape'
+  uint32_t new_shape_idx = 0;
+  while (new_shape_idx < new_shape.size())
+  {
+    Dim &target_dim = new_shape[new_shape_idx];
+
+    // Ignore dim == 1
+    if (target_dim.value == 1)
+    {
+      new_shape_idx++;
+      continue;
+    }
+
+    while (old_shape_product < target_dim.value)
+    {
+      if (expand_range() == false)
+        break;
+    }
+
+    if (old_shape_product != target_dim.value)
+      return false;
+
+    // old_shape_product == target_dim.value
+    for (uint32_t idx = old_shape_start_idx; idx < old_shape_end_idx; idx++)
+    {
+      const auto &old_tags = _shape[idx].tags;
+      target_dim.tags.insert(target_dim.tags.end(), old_tags.begin(), old_tags.end());
+    }
+
+    new_shape_idx++;
+    move_to_next_range();
+  }
+  _shape = new_shape;
+  return true;
+}
+
+bool TaggedShapeAnalyzer::verify_tag() const
+{
+  // check whether tags in _shape are incremental
+  uint8_t tag = START_TAG;
+  for (const auto &dim : _shape)
+  {
+    for (const auto &t : dim.tags)
+    {
+      if (t == tag)
+        tag++;
+      else
+        return false;
+    }
+  }
+  return true;
+}
+
+template <loco::DataType DType>
```

Comment:
Why did you use template? It seems that perm and shape are always int32.

Reply:
At first, I didn't know that `perm` and `shape` are always given as `int32`.
So, I used a template to support other types. 

For now, I just leave it just in case (maybe can be used later).

Reply:
I'm fine to remove a template (It might hurt readability a bit). 
Do you prefer to remove it?

Reply:
That's ok. Let's keep it.